pkg/encode/encoder.go

```diff
@@ -0,0 +1,431 @@
+package encode
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	commons "github.com/kitex-contrib/codec-hessian2/pkg/common"
+	"io"
+	"math"
+	"reflect"
+	"strings"
+)
+
+func NewHessian2Output(w io.Writer) *Encoder {
+	return &Encoder{
+		writer: w,
+		buffer: bytes.NewBuffer(nil),
+	}
+}
+
+// Class表示Java对象的类型信息
+type Class struct {
+	Name   string   // 类名
+	Fields []*Field // 字段信息
+}
+
+// Field表示Java对象的字段信息
+type Field struct {
+	Name    string // 字段名
+	Type    *Class // 字段类型
+	Default Object // 字段默认值
+}
+
+// Object表示Java对象的值信息
+type Object interface{}
+
+type Encoder struct {
+	writer io.Writer     // 输出流
+	buffer *bytes.Buffer // buffer缓存
+
+	writeType      bool     // 是否写入类型信息
+	typeRefWritten bool     // 是否已经将类型引用写入
+	typeRefs       []*Class // 类型引用数组
+	objRefs        []Object // 对象引用数组
+}
+
+func (e *Encoder) writeObjectRef(i int) error {
+	if err := e.WriteByte('R'); err != nil {
+		return err
+	}
+	if err := binary.Write(e.buffer, binary.BigEndian, i); err != nil {
+		return err
+	}
+	return nil
+}
+func (e *Encoder) WriteObject(obj Object) error {
+	// 判断是否为nil
+	if obj == nil {
+		return e.WriteNull()
+	}
+
+	// 获取对象类型信息
+	clazz, err := getClass(obj)
+	if err != nil {
+		return err
+	}
+
+	// 写入类型信息
+	if e.writeType {
+		if err := e.writeClass(clazz); err != nil {
+			return err
+		}
+	}
+
+	// 写入对象引用
+	if ref, ok := e.findObjRef(obj); ok {
+		return e.writeObjectRef(ref)
+	}
+
+	// 将对象添加到引用数组
+	e.objRefs = append(e.objRefs, obj)
+
+	// 判断类型并编码
+	switch v := obj.(type) {
+	case bool:
+		return e.WriteBool(v)
+	case int:
+		return e.WriteInt(int64(v))
+	case int8:
+		return e.WriteInt(int64(v))
+	case int16:
+		return e.WriteInt(int64(v))
+	case int32:
+		return e.WriteInt(int64(v))
+	case int64:
+		return e.WriteInt(v)
+	case uint:
+		return e.WriteInt(int64(v))
+	case uint8:
+		return e.WriteInt(int64(v))
+	case uint16:
+		return e.WriteInt(int64(v))
+	case uint32:
+		return e.WriteInt(int64(v))
+	case uint64:
+		return e.WriteInt(int64(v))
+	case float32:
+		return e.writeDouble(float64(v))
+	case float64:
+		return e.writeDouble(v)
+	case string:
+		return e.WriteString(v)
+	}
+
+	// 编码复杂类型
+	switch v := obj.(type) {
+	case []byte:
+		return e.writeBytes(v)
+	case []interface{}:
+		return e.writeList(v)
+	case map[string]interface{}:
+		return e.writeMap(v)
+	}
+
+	// 构造并编码Java对象
+	return e.writeClass(clazz)
```

Comment:
only class info being wrote, obj instance data lost

Reply:
the data was written before,this is wirte class type info